clients/src/main/java/org/apache/kafka/common/protocol/Errors.java

```diff
@@ -362,7 +363,8 @@
     BROKER_ID_NOT_REGISTERED(102, "The given broker ID was not registered.", BrokerIdNotRegisteredException::new),
     INCONSISTENT_TOPIC_ID(103, "The log's topic ID did not match the topic ID in the request", InconsistentTopicIdException::new),
     INCONSISTENT_CLUSTER_ID(104, "The clusterId in the request does not match that found on the server", InconsistentClusterIdException::new),
-    TRANSACTIONAL_ID_NOT_FOUND(105, "The transactionalId could not be found", TransactionalIdNotFoundException::new);
+    TRANSACTIONAL_ID_NOT_FOUND(105, "The transactionalId could not be found", TransactionalIdNotFoundException::new),
+    OFFSET_MOVED_TO_TIERED_STORAGE(107, "The requested offset is moved to tiered storage.", OffsetMovedToTieredStorageException::new);
```

Comment:
Copied this error number `107` from https://github.com/apache/kafka/pull/11390

Reply:
Any thoughts on how we would upgrade/downgrade a cluster which has changes to error enums, but which will potentially conflict with error enums in some future, upstream release?

Reply:
I agree. If a new Error enum is introduced before the PR mentioned above gets merged, and gets checked in with id `107`, we will be in trouble when upgrading.

One way out is to use a large error ID like 1001 for now, until the code gets merged.

Reply:
If we add it with error code 1000 and give it a custom name, we can deprecate it's use later when the error is available in upstream.

For a while, we will have the functionality to process both errors, but emit only the upstream error. Then, with the next rolling update, we can remove it and just use the error from upstream.

Does that sound reasonable?

Reply:
That sounds good.  What about other places where we have magic version numbers like the various message versions that are created?

Reply:
One place I can think of is the timestamp `-4` used to signify `EARLIEST_LOCAL_TIMESTAMP` in a `ListOffsetsRequest`. We can probably change this to `-101`, and then later start treating both `-4` and `-101` the same way.

Reply:
I'm ok with those solutions.  @gitlw?

Reply:
Actually, do we need to make a change ApiVersion.scala?

Reply:
In this PR, the EARLIEST_LOCAL_TIMESTAMP enum is not exposed to the client as the client is using https://github.com/linkedin/kafka/blob/bd72ef1bf1e40feb3bc17349a385b479fa5fa530/clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java#L24 in the KafkaAdminClient.listOffsets API.
I'm not sure if this enum will be exposed to the client in a subsequent PR or in upstream kafka. If so, it'll be harder to deprecate.
If not, then I'm good with the suggested approach above.

Reply:
I don't know why clients would ever use this request.  It seems like it's just issued by brokers.

Reply:
We will not expose this to clients as -101. If it exposed to clients in upstream Kafka, it will be with the code -4 or -5. We can still deprecate -101 without disruption.

Reply:
Deprecating a change in clients involves chasing who is using this, and may sometimes turn out to be a full-blown HI.
Fortunately, it's not exposed to clients now, and thus we are good.